File: backend/src/main/java/com/nak/backend/erp/hr/service/DepartmentService.java
package com.nak.backend.erp.hr.service;

import com.nak.backend.erp.hr.dto.DepartmentDto;
import com.nak.backend.erp.hr.mapper.DepartmentMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;
import java.util.*;
import java.util.stream.Collectors;

/**
 * ERP 부서 Service
 * 부서 데이터를 트리 구조로 제공하며, 하위 부서가 있는 부서는 삭제되지 않도록 보호합니다.
 */
@Service
@RequiredArgsConstructor
public class DepartmentService {

    private final DepartmentMapper departmentMapper;

    /**
     * 모든 부서를 트리 구조로 조회합니다.
     */
    public List<DepartmentDto> getDepartmentTree() {
        List<DepartmentDto> allDepts = departmentMapper.findAll();
        return buildTree(allDepts);
    }

    /**
     * 모든 부서를 평면 리스트로 조회합니다.
     */
    public List<DepartmentDto> getAllDepartments() {
        return departmentMapper.findAll();
    }

    /**
     * 부서 단건 조회
     */
    public DepartmentDto getDepartment(String id) {
        DepartmentDto dept = departmentMapper.findById(id);
        if (dept == null) {
            throw new NoSuchElementException("Department not found: " + id);
        }
        return dept;
    }

    /**
     * 부서 생성
     */
    @Transactional
    public DepartmentDto createDepartment(DepartmentDto dto) {
        String id = UUID.randomUUID().toString();
        dto.setId(id);
        if (dto.getSortOrder() == null) dto.setSortOrder(0);
        if (dto.getIsActive() == null) dto.setIsActive(true);
        dto.setCreatedAt(OffsetDateTime.now());
        dto.setUpdatedAt(OffsetDateTime.now());

        departmentMapper.insert(dto);
        return dto;
    }

    /**
     * 부서 수정
     */
    @Transactional
    public DepartmentDto updateDepartment(String id, DepartmentDto dto) {
        DepartmentDto existing = departmentMapper.findById(id);
        if (existing == null) {
            throw new NoSuchElementException("Department not found: " + id);
        }

        dto.setId(id);
        dto.setCreatedAt(existing.getCreatedAt());
        dto.setUpdatedAt(OffsetDateTime.now());
        departmentMapper.update(dto);
        return dto;
    }

    /**
     * 부서 삭제 (하위 부서가 있으면 삭제 불가)
     */
    @Transactional
    public void deleteDepartment(String id) {
        DepartmentDto dept = departmentMapper.findById(id);
        if (dept == null) {
            throw new NoSuchElementException("Department not found: " + id);
        }

        List<DepartmentDto> children = departmentMapper.findByParentId(id);
        if (!children.isEmpty()) {
            throw new IllegalStateException("Cannot delete department with child departments. Delete children first.");
        }

        departmentMapper.deleteById(id);
    }

    /**
     * 평면 리스트를 트리 구조로 변환
     */
    private List<DepartmentDto> buildTree(List<DepartmentDto> flatList) {
        if (flatList == null || flatList.isEmpty()) return Collections.emptyList();

        Map<String, List<DepartmentDto>> parentChildMap = flatList.stream()
                .collect(Collectors.groupingBy(
                        d -> d.getParentId() != null ? d.getParentId() : "",
                        Collectors.toList()
                ));

        // 루트 노드 (parentId가 null인 노드)
        List<DepartmentDto> roots = flatList.stream()
                .filter(d -> d.getParentId() == null)
                .collect(Collectors.toList());

        // 각 노드에 children 할당
        for (DepartmentDto dept : flatList) {
            String deptId = dept.getId();
            List<DepartmentDto> children = parentChildMap.getOrDefault(deptId, Collections.emptyList());
            if (!children.isEmpty()) {
                children.sort(Comparator.comparingInt(DepartmentDto::getSortOrder));
                dept.setChildren(children);
            }
        }

        roots.sort(Comparator.comparingInt(DepartmentDto::getSortOrder));
        return roots;
    }
}

File: nodebackend/src/domains/department/department.service.ts
import { Injectable, NotFoundException, ConflictException } from '@nestjs/common';
import { PrismaService } from '../../prisma.service';

@Injectable()
export class DepartmentService {
  constructor(private readonly prisma: PrismaService) {}

  private mapDepartment(dept: any) {
    return {
      id: dept.id,
      parentId: dept.parent_id,
      name: dept.name,
      code: dept.code,
      managerId: dept.manager_id,
      sortOrder: Number(dept.sort_order || 0),
      isActive: dept.is_active,
      createdAt: dept.created_at,
      updatedAt: dept.updated_at,
    };
  }

  async getDepartments() {
    const depts = await this.prisma.erp_departments.findMany({
      orderBy: { created_at: 'asc' }
    });
    return depts.map(d => this.mapDepartment(d));
  }

  async getDepartmentTree() {
    const depts = await this.prisma.erp_departments.findMany({
      orderBy: { created_at: 'asc' }
    });

    const map = new Map();
    const roots: any[] = [];

    depts.forEach(d => {
      map.set(d.id, { ...this.mapDepartment(d), children: [] });
    });

    depts.forEach(d => {
      if (d.parent_id && map.has(d.parent_id)) {
        map.get(d.parent_id).children.push(map.get(d.id));
      } else {
        roots.push(map.get(d.id));
      }
    });

    return roots;
  }

  async getDepartment(id: string) {
    const dept = await this.prisma.erp_departments.findUnique({ where: { id } });
    if (!dept) throw new NotFoundException(`Department not found: ${id}`);
    return this.mapDepartment(dept);
  }

  async createDepartment(dto: any) {
    const created = await this.prisma.erp_departments.create({
      data: {
        parent_id: dto.parentId || null,
        name: dto.name,
        code: dto.code,
        manager_id: dto.managerId || null,
        sort_order: dto.sortOrder ?? 0,
        is_active: dto.isActive ?? true,
      }
    });
    return this.mapDepartment(created);
  }

  async updateDepartment(id: string, dto: any) {
    const existing = await this.prisma.erp_departments.findUnique({ where: { id } });
    if (!existing) throw new NotFoundException(`Department not found: ${id}`);

    const updated = await this.prisma.erp_departments.update({
      where: { id },
      data: {
        parent_id: dto.parentId,
        name: dto.name,
        code: dto.code,
        manager_id: dto.managerId,
        sort_order: dto.sortOrder,
        is_active: dto.isActive,
      }
    });
    return this.mapDepartment(updated);
  }

  async deleteDepartment(id: string) {
    const existing = await this.prisma.erp_departments.findUnique({ where: { id } });
    if (!existing) throw new NotFoundException(`Department not found: ${id}`);

    // 하위 부서가 있으면 삭제 불가 (backend 동일 로직)
    const children = await this.prisma.erp_departments.findMany({
      where: { parent_id: id }
    });
    if (children.length > 0) {
      throw new ConflictException('Cannot delete department with child departments. Delete children first.');
    }

    await this.prisma.erp_departments.delete({ where: { id } });
  }
}
